### ISSUE TITLE
How can I adjust the load on CPU

### ISSUE THREAD

**Issue by podaen** [bertmelis/espMqttClient#29]
I want to adjust the load on the CPU.  I am focusing on...

> Publishing at QoS 0, packetId: 1

I think this causes a delay in my soft. I want to begin with shorten the time that it's occupied with it? It's not a big message but maybe I can do something to improve it?

esp32
Arduino IDE 1.8.16
Arduino ESP 2.0.4

**Comment by Pablo2048** [comment]
What tool/code do you use to measure CPU load by espMqtt task? Did you try to change the task priority?

**Comment by podaen** [comment]
I am using my own sketch. Where can I specify it? I checked the Config.h, but it's not there.

**Comment by Pablo2048** [comment]
https://github.com/bertmelis/espMqttClient/blob/1133c34d245350330b59d4af553570bc202cdc35/src/MqttClient.cpp#L17

**Comment by bertmelis** [comment]
You're using a microcontroller. Load on the CPU is Avery abstract thing here.

I did a test yesterday and sent 100 qos2 messages in less than a few seconds. Don't know how fast you want it to be but as @Pablo2048 said, on esp32 you can set a task priority (and core to run on).

**Comment by podaen** [comment]
I do. What's the difference with qos0, qos1, qo2? A few secondes is just to slow.  I am only publishing 3 characters. The priority on my task is 2.

> https://github.com/bertmelis/espMqttClient/blob/1133c34d245350330b59d4af553570bc202cdc35/src/MqttClient.cpp#L17

On what priority is it initialized to?

**Comment by Pablo2048** [comment]
QOS are specified in MQTT protocol specification - its easier if you find and read it on the internet. Where you get "few seconds" delay? The publish function just adds the message to the queue and quit - see https://github.com/bertmelis/espMqttClient/blob/1133c34d245350330b59d4af553570bc202cdc35/src/MqttClient.cpp#L163 . Default priority is 1 - see https://github.com/bertmelis/espMqttClient/blob/1133c34d245350330b59d4af553570bc202cdc35/src/espMqttClient.h#L22

**Comment by podaen** [comment]
> Where you get "few seconds" delay?

Sorry, I miss understood. I will try to set the priority.

**Comment by podaen** [comment]
Still have it in the beginning, but it's getting beter. 

Don't know how you do it, but I did it once for a keypad like this.

```
void Task (void* pvParameters) {

	TickType_t xLastWakeTime;
	const TickType_t xFrequency = 200;// Perform an action every 10 ticks.//10//1000=evry second
	xLastWakeTime = xTaskGetTickCount();// Initialise the xLastWakeTime variable with the current time.

	for (;;) {//infinit loop
		delay(5);
		vTaskDelayUntil(&xLastWakeTime, xFrequency);// Wait for the next cycle.
		//Serial.println("Do something");
	}
}
```

It just wait until it done some cycles to do something.

**Comment by podaen** [comment]
How do I know if I got enough memory on the task? It seems to adjust to value I set it to.

> [  7469][I][MqttClient.cpp:263] loop(): Stack usage: 9444/10000

> [  7534][I][MqttClient.cpp:263] loop(): Stack usage: 14448/15000

```
#ifndef EMC_TASK_STACK_SIZE
#define EMC_TASK_STACK_SIZE 15000
#endif
```

**Comment by bertmelis** [comment]
That log statement is a bit misleading. It shows the highwatermark for the MQTT task compared to the total amount of stack memory: https://github.com/bertmelis/espMqttClient/blob/main/src/MqttClient.cpp#L263

So you used 556 out of 10000 **words**. See https://www.freertos.org/uxTaskGetStackHighWaterMark.html
Meaning, you have plenty of headspace and it is safe to lower the value.

**Comment by podaen** [comment]
I see you are using xSemaphore, you can add vTaskDelayUntil to give main tasks more time to do things.

https://stackoverflow.com/questions/41652270/freertos-vtaskdelayuntil-finishes-immediately

**Comment by Pablo2048** [comment]
It makes no sense - `vTaskDelayUntil` return immediately only if there is no time (or even "negative" time) to wait for. The used FreeRTOS is preemptive and espMqttClient task priority is just above task idle. Can you please explain more, what exactly you are trying to achieve and put here MCVE, which can be tested/worked on?

**Comment by bertmelis** [comment]
The MQTT task yields here: https://github.com/bertmelis/espMqttClient/blob/main/src/MqttClient.cpp#L258

Expands to `taskYIELD()` which is the efficient equivalent of `vTaskDelay(0)`.

As @Pablo2048 said: please describe the problem you have, provide an example and [not the desired solution](https://xyproblem.info/).

**Comment by podaen** [comment]
What do you mean by "no time (or even "negative" time) to wait for"? The vTaskDelayUntil doesn't wait on timebase. It says delay, but it doesn't delay anything. It just count the ticks, and if the value has reached it will continue. Depending on the xFrequency of the CPU it runs over a the loop, with this you can adjust how many times it runs over the code in the same time frame.

I am running streams that need lots memory and cpu power. If I don't be there in time it bangs in my speaker and this is caused because the cpu is in use by an other task. I had this problem before with an I2C device.

In MqttClient::loop(), I see that it's doing all these things. `_checkOutgoing();` , `_checkIncoming();` , `_checkPing();` and I when I starts to publishing it is just to much.

**Comment by bertmelis** [comment]
Raise the priority of the critical task, work with critical sections, change the core it runs on. FreeRTOS gives you many tools.

I'm still not convinced to change anything. Are you using the sync or the async version? Did you profile the code? What do you propose to do? An MCVE?

**Comment by podaen** [comment]
Priority has all ready been set and it improved but the problem is still there. Don't know if I can do something with the FreeRTOS tools because the semaphore gives and takes is there... Any suggestions?

I am using the encrypted example, I didn't change anything except set the priority of the mqtt client to 2 and added my task with priority 1. Don't know if it's sync or async. I had to `#include <AsyncTCP.h>` so I assume it is async...

I will look if I can show it to you in code. It's not an easy job you know.

**Comment by bertmelis** [comment]
- If your task is time critical, why do you give it a lower priority? Higher = more CPU time.
- This lib includes AsyncTCP.h so you shouldn't need to do it yourself. Maybe something else in your code needs it
- Encryption eats a lot of cpu cycles. Maybe raise the cpu frequency?

**Comment by podaen** [comment]
I just checked and there is no other task running than these.

```
		xTaskCreatePinnedToCore(
			Task1,   /* Task function. */
			"URL1",     /* name of task. */
			STACK_SIZE,       /* Stack size of task */
			NULL,        /* parameter of the task */
			1,           /* priority of the task */
			&URL1,  /* Task handle. */
			1); 
		delay(500);
```

`espMqttClientSecure mqttClient(2, 1);`

The CPU fequentie is set to full speed. I just tried to set it to core 0... Now it's waiting.

Why would you not add the possibility to set the xFrequency to run the loop? Just disable it and with define it can be customized

**Comment by bertmelis** [comment]
Best I can do is (optionally) disabling the automatic task creation (as done in the async version) so you can call `loop()` as you like.

Actually, the async version does exactly this by calling `loop()` every 500msec via async's `onPoll`.

I'm doing the same for the Linux version that I'm testing where I create a thread in the main program that calls `loop()`. (see Linux branch).

Let me think about this -how to implement- because I don't want to create too many different flavours which all bring in a bit bloat into the lib.

**Comment by podaen** [comment]
Now my program is finished and evryting is working fine. I did simplefied my other tasks and set the core and priority. For me you don't have to do something Additionly.

**Comment by podaen** [comment]
How do I use only the loop?

**Comment by bertmelis** [comment]
Haven't implemented this yet.

**Comment by podaen** [comment]
Don't mind. I don't think the problem is in here. But all my clients are disconnecting, for the mqtt it is `stop_ssl_socket()` and eventually the WiFi disconnects. I think it is because the want to read the buffer on the same time. I'm not sure...

**Comment by podaen** [comment]
I found it! I set the WiFi power management to maximum power `WiFi.setSleep(false);`. I don't have any failure anymore or a slow connection. It surely improved working behavior and I am glad that I finally found some improvement. I search a long time for this and don't understand this is a default setting in the SDK of ESP Arduino. A battery is optional... and it is not that I am overclocking.

**Comment by podaen** [comment]
The WiFi.setSleep(false); did improved it, but is was not over yet. So I search for it, what could be the problem and now I found it. The esp had some powering suffering issues, but just enough to work. I solved that issues and one by one all my problems disappear.

**Comment by bertmelis** [comment]
It can be quite power hungry when a lot of WiFi traffic is going on.
